core/arch/x86/encode.c

```diff
@@ -2479,7 +2529,8 @@ instr_encode_arch(dcontext_t *dcontext, instr_t *instr, byte *copy_pc, byte *fin
         encode_operand(&di, info->src2_type, info->src2_size, instr_get_src(instr, 1));
     if (info->src3_type != TYPE_NONE)
         encode_operand(&di, info->src3_type, info->src3_size, instr_get_src(instr, 2));
-    if (di.mod == 5 && di.reg < 8) { /* mod may never be set (e.g., OP_extrq) */
+    if (!opc_is_opmask_arch(di.opcode) && di.mod == 5 &&
```

Comment:
I think we'd rather use a bit or flag inside info rather than checking a list of opcodes.

Reply:
There are a few uses alike like instr_is_cti, instr_get_predicate etc. Before I do change it, how exactly? You mean flag in the decoder table (nah..)? Or do you mean just a bool flag like bool is_opmask = opc_is_opmask_arch(di.opcode) further up? If not a flag in the table, which I think is unnecessary, then one has to check the opcode _somewhere_.

Reply:
A flag in the table.  Why have an explicit series of branches looking at opcodes when we have per-opcode info right there in the table entry?

Reply:
> There are a few uses alike like instr_is_cti, instr_get_predicate etc.

During encoding?  I wouldn't think so: the encoder should not go and query things externally based on the opcode.  The encoding table entry should have everything it needs.  It should not be dispatching on opcode: that's the whole point of the encoding table, to have everything in one spot.

Reply:
Good that I (had to) look at this again. I had fixed a bug (in my patch) to support TYPE_K_MODRM k registers OR memory. The memory part through encoding modrm was missing. Now modrm.mod will always be set by the operands, the special opmask case here is not needed anymore. I removed it, including the functions I had added. Should be clean now.